python/ray/_raylet.pyx

```diff
@@ -404,6 +417,259 @@ cdef class RayletClient:
+
+            backtrace = ray.utils.format_error_message(
+                traceback.format_exc(), task_exception=task_exception)
+            if isinstance(error, RayTaskError):
+                # Avoid recursive nesting of RayTaskError.
+                failure_object = RayTaskError(function_name, backtrace,
+                                              error.cause_cls)
+            else:
+                failure_object = RayTaskError(function_name, backtrace,
+                                              error.__class__)
+            _store_task_outputs(
+                worker, return_ids, [failure_object] * len(return_ids))
+            ray.utils.push_error_to_driver(
+                worker,
+                ray_constants.TASK_PUSH_ERROR,
+                str(failure_object),
+                job_id=worker.current_job_id)
+
+            # Send signal with the error.
+            ray_signal.send(ray_signal.ErrorSignal(str(failure_object)))
+
+    # Reset the state fields so the next task can run.
+    worker.task_context.current_task_id = TaskID.nil()
+    worker.core_worker.set_current_task_id(TaskID.nil())
+    worker.task_context.task_index = 0
+    worker.task_context.put_index = 1
+
+    # Don't need to reset `current_job_id` if the worker is an
+    # actor. Because the following tasks should all have the
+    # same driver id.
+    if <int>task_type == <int>TASK_TYPE_NORMAL_TASK:
+        worker.current_job_id = JobID.nil()
+        worker.core_worker.set_current_job_id(JobID.nil())
+
+        # Reset signal counters so that the next task can get
+        # all past signals.
+        ray_signal.reset()
+
+    # Reset the state of the worker for the next task to execute.
+    # Increase the task execution counter.
+    worker.function_actor_manager.increase_task_counter(
+        job_id, function_descriptor)
+
+    # If we've reached the max number of executions for this worker, exit.
+    reached_max_executions = (
+        worker.function_actor_manager.get_task_counter(
+            job_id, function_descriptor) == execution_info.max_calls)
+    if reached_max_executions:
+        worker.core_worker.disconnect()
+        sys.exit(0)
+
+cdef CRayStatus execute_task_callback(
+        CTaskType task_type,
+        const CRayFunction &ray_function,
+        const CActorID &c_actor_id,
+        const unordered_map[c_string, double] &c_resources,
+        const c_vector[shared_ptr[CRayObject]] &c_args,
+        const c_vector[CObjectID] &c_arg_reference_ids,
+        const c_vector[CObjectID] &c_return_ids,
+        c_vector[shared_ptr[CRayObject]] *returns) nogil:
+
+    with gil:
+        try:
+            # The call to execute_task should never raise an exception. If it
+            # does, that indicates that there was an unexpected internal error.
+            execute_task(task_type, ray_function, c_actor_id, c_resources,
+                         c_args, c_arg_reference_ids, c_return_ids, returns)
```

Comment:
I think we will want to return a not OK status if the task threw an exception so that the core worker can handle errors.

Reply:
Agreed, but that will require more changes. Plan to do it in a follow-up. For now, I'm just replicating the code that was there before (If there was an internal error, exit. If there was a task error, it puts an exception in the store and returns "success").